### ISSUE TITLE
QPrompt AppImage requires GLIBC_2.29, making it incompatible with Ubuntu 18.04 based distros

### ISSUE THREAD

**Issue by leirbag28** [Cuperino/QPrompt#31]
**Bug Description**
This Applmage is not universal, it only works on Ubuntu 20.04 (and clones) or higher as it requires GLIBC_2.29 so does not work on Ubuntu 18.04 (and clones) and that is quite a lot as Ubuntu 18.04 is still heavily used.

**Steps to Reproduce**

1. Drop QPrompt AppImage into Ubuntu MATE terminal and it shows what the problem is... which is it requires GLIBC_2.29

**Expected behavior**
For it to work.

**Device information**
 - Operating System: Ubuntu MATE 18.04.5 64bit
 - Hardware: Acer Chromebook C720p 4gb RAM
 - QPrompt Version: v1.0.0-2c9752c-linux-64-gcc AppImage

**Additional context**
Perhaps it can be made to NOT require GLIBC_2.29 like how Imaginary Teleprompter does not require it?

**Comment by Cuperino** [comment]
I thought this only affected the Deb package... (available for arm64 only, as I'm trying to stay away from builds that are not universal)

The high version requirement of GLIBC is a by-product of the system from which QPrompt is compiled. I will see if I can force an older version of GLIBC to build the AppImage.

I cannot release a native build for Ubuntu 18.04 because the version of Qt in its repositories is too old to run QPrompt. QPrompt can be modified and built to run there, but the user experience is not the one I intend to give.

Speaking of universal packages, QPrompt is also available as a Snap and a Flatpak. The Snap version has been proven to work on 18.04; it uses the 20.04 core from the Snap environment.

The Flatpak arrived a few hours ago, and I've only tested it on newer systems. If you'd be able and willing to test the Flatpak and report on it I'd be very grateful:

https://flathub.org/apps/details/com.cuperino.qprompt

**Comment by leirbag28** [comment]
Thanks Javier! Superb app and support, i mean you guys have a SNAP, Flatpak, Appimage and more. Amazing!

I DID install the SNAP and it works superb...I prefered a Flatpakbut there wasnt any...lo and behold like a day later you guys posted one haha. So, will having the SNAP version and Flatpak version conflict with each other?

Anyhow on the original subject, you can make an Appimage without GLIBC_2.29 without it necessarily being an ubuntu 18.04 app but rather just an Appimage that should work on all systems including ubuntu 16.04 and higher. Maybe use QT4 or lower? I dunno. 

Also one or two major requests. Can i add an optiion to remove the controls (the blue circle with return arrow and two little arrows on the side on the bottom center screen) for the MAIN screen (I already know its not on the external screen monitor) and also to add blurry shadow to the text? 
Why you ask? It will add Tons more usage to this app! I use it to sing Lyrics while showing my face on video in the background and record the entire output. Also as and End Credits for a movie or video. Also can be used for  a public Speaker to actually show the words for the audience to see (like a powerpoint) but still showing video of the speaker. I make the text transparent and smaller and align it to the left or right, and on the clean side my video of me showing clearly while the text scrolls next to me. Also I am experimenting with Tik Tok videos having what I want to say scrolling over my face and screen capturing my face with SimpleScreenRecorder grabbing the vertical rectangle area or aspect ratio for tiktok.

Your app is amazing! And.....on Linux!! Thank you.

**Comment by Cuperino** [comment]
Thank you! Glad you like it!

I'm a Linux user so, QPrompt had to be a Linux app that also worked on other platforms. It's written using Qt and KDE's Kirigami framework for the UI. 

Qt 5 can work on older systems if it's properly containerized. I think 16.04 is the oldest QPrompt could go because prior to that universal packages did not exist, Snaps were the first and those were invented in 2015. I doubt it's possible to make the AppImage work on on 16.04 because QPrompt requires the latest Qt 5.15 to work, and I doubt Qt 5.15 can be compiled with a version of GLIBC so old that it works on 16.04.

While I love older systems, and support some, I'd like to focus on developing for the future and not for the past. Trust me that if time wasn't limited I would write a teleprompter program for the SEGA Genesis in assembly. But there are more important things to work on and many apps that don't yet exist or don't come for Linux.

> Can i add an option to remove the controls (the blue circle with return arrow and two little arrows on the side on the bottom center screen) for the MAIN screen

The arrows you can hide by changing Pointers to None.

I worry that hiding the option to return to edit mode would cause some users to panic... I could have it not appear at the bottom and have the top bar not disappear while in full screen instead. Unfortunately, it is not possible to completely hide the option because it is managed by the Kirigami framework and I'm avoiding at all cost making modifications to it, unless those modifications are upstream bug fixes. I don't want to modify Kirigami because then it would be another thing that I'd have to maintain.

> Also as and End Credits for a movie or video.

This should be done using a video editor or a dedicated character generator... Users are free to use QPrompt for this purpose, but I won't support it.

> Also I am experimenting with Tik Tok videos having what I want to say scrolling over my face and screen capturing my face with SimpleScreenRecorder grabbing the vertical rectangle area or aspect ratio for tiktok

SimpleScreenRecorder works great, but you should consider a dedicated capture card for this as it'll help prevent frame drops. You could then crop the video using ffmpeg or DaVinci Resolve, so you have no quality loss. I don't recommend using Kdenlive or Pitivi for cropping because their crops affect the color space.

> and also to add blurry shadow to the text?

This is an interesting proposition. A shader could be used to show shadows behind the contents and it'll help improve readability with a fully transparent background. I'll add this to the to do list.

Whenever you find a bug or think of a new feature to add, please file a new issue. It helps keep things organized. Thank you for your great remarks and suggestions. This one here is truly excellent!

**Comment by Cuperino** [comment]
>  So, will having the SNAP version and Flatpak version conflict with each other?

There's no conflict between any of the available install options. Of course, if you attempt to launch from the terminal, which version is launched will be determined by your shell's path. You can run `whereis qprompt` to see which version will be the first to launch.

**Comment by leirbag28** [comment]
Just tried the Flatpak version along with the already installed Snap version on Ubuntu 18.04 MATE and Compton compositor.

The Flatpak works fine just as the SNAP. However I noticed both cannot get out of FULLSCREEN (not the Fullscreen withing the app) by pressing F11 on the keyboard...so i cant move the window. Have to press ALT-Tab to see whats behind or switch to whats behind.

Also are you aware there are 2 things already named Qprompt? One is a teleprompter the other a python library:

Other QPrompts:
https://qprompt.com/

https://pypi.org/project/qprompt/

**Comment by Cuperino** [comment]
Good to know it works with Compton. I tried making an AppImage for 18.04 but didn't have success with it. The older AppImage builder can't parse some of the newer tags used by the Freedesktop standards today and I'm not sure removing those and maintaining separate manifest files is a good idea, since 

With regards to the other marks out there, those are stylized different from QPrompt. _Qprompt_ is a library for developing quick CLI user prompts, input, and menus, and _Q-Prompt Inc._ is a services company offering teleprompter solutions. _QPrrompt App_ is not related to Q-Prompt Inc. in any way.

The reasons why QPrompt was chosen are written in the [Pun Facts](https://github.com/Cuperino/QPrompt#pun-facts) section of QPrompt's readme. I will add that QPrompt is not related to Q-Prompt Inc. to avoid confusion.

**Comment by leirbag28** [comment]
What if you included the required libraries ir dependencies that Ubuntu 18.04 lacked right into the Appimage? Because the flatpak and snap work in 18.04?
I personally think ultimately AppImage is the best of the 3. I can install 1000 apps by just dragging them into a customers linux pc, and delete all 1000 just as fast. No knowledge whatsoever required. Snaps have partially wrecked my system, so i go for flatpaks first if available. Its a good thing they exist for bow but I hope AppImage wins this war.

QPrompt name:

Just thought i'd point it out that those others exist...will certainly confuse people and make it harder to find your app and harder for the name to stand out.

Incase you wantes to rename:

Q-Teleprompter
Prompter-Q
Teleprompter-Q
Kyew Prompter
Qyoo Promter
Cue-Prompt
Cutie-Prompter
Cute-Prompter (like CutefishOS which has built in QT and was named like that for that reason)

**Comment by Cuperino** [comment]
I've successfully created an AppImage compatible with Ubuntu 18.04. This required me to drop a series of metadata attributes needed to take advantage of the features in modern Linux software stores. Some of those features are absolutely essential to support mobile Linux distributions.

I think current mobile distributions should take precedence over distributions which are about to drop support a year from now. Nevertheless, there currently seem to be more users running QPrompt in these older distros than there are installing the arm64 Flatpak; meaning there are more users in these older distros than there are in mobile phones and tablets that run Linux. For this reason, the next couple of releases should support Ubuntu 18.04 and derivative distributions in their AppImage builds.

Support for Ubuntu 18.04 may be limited to releases made in 2022. Some betas may not be included among those releases because supporting the 18.04 base requires me to use a VM to make the builds, plus maintain a separate code branch free of modern AppData metadata attributes.

**Comment by Cuperino** [comment]
I've uploaded an AppImage of the most current development build, made on Ubuntu 18.04 for best compatibility.

@leirbag28, would you be so kind to test it and indicate how well it performs compared to the Snap and the Flatpak?
https://github.com/Cuperino/QPrompt/releases/tag/v1.1.0-beta-1-AppImage

**Comment by Cuperino** [comment]
It's official. I'm sticking to AppImages that can work on Ubuntu 18.04 for as long as KDE's Craft build pipeline allows me to. These are being built on an Ubuntu 18.04 VM.

**Comment by leirbag28** [comment]
Well that's freakin cool. I was just defaulting to the AppImage version two days ago because the Flatpak version Injust updated, didnt allow me to show a second display...I played around with the SNAP and Appimage and they seemed to do the same thing, then I just logged out and back in and that seemed to fix the issue.

**Comment by Cuperino** [comment]
Sounds like a system component just needed to re-launched after an update. This is necessary because all programs must be copied to RAM in order to be used. That is at least the case for most digital computers and is done this way for performance reasons. Logging out and back in usually does the trick for GUI programs, unless it's something critical like the display server itself or the login manager, then it's trickier to re-launch without restarting. Not recommended, but still possible, at least on Linux.

FYI, QPrompt's Flatpak is using an older runtime because of a bug in that was in the most current runtime last time we tried to update to it. Hopefully it should have been fixed by now and the next stable update should be able to use it without an issue.

With regards to AppImage, QPrompt 1.1.0 betas 1 through 4 should work with Ubuntu 18.04. Those can be found at GitHub releases: https://github.com/Cuperino/QPrompt/releases
